spring-cloud-azure-config/src/main/java/com/microsoft/azure/spring/cloud/config/msi/ConfigMSICredentials.java

```diff
@@ -0,0 +1,196 @@
+                }
+            } finally {
+                if (response != null) {
+                    try {
+                        response.close();
+                    } catch (IOException e) {
+                        LOGGER.warn("Failed to close the response", e);
+                    }
+                }
+            }
+        }
+
+        throw new IllegalStateException(RETRIEVE_IMDS_FAILURE_MSG);
+    }
+
+    private String toString(CloseableHttpResponse response) {
+        StringWriter writer = new StringWriter();
+        try {
+            IOUtils.copy(response.getEntity().getContent(), writer, "UTF-8");
+        } catch (IOException e) {
+            LOGGER.warn("Failed to convert response {} to String.", response, e);
+        }
+        return writer.toString();
+    }
+
+    private boolean canRetry(CloseableHttpResponse response) {
+        if (response == null) {
+            return false;
+        }
+
+        int statusCode = response.getStatusLine().getStatusCode();
+        return statusCode == 404 || statusCode == 429 || (statusCode >= 500 && statusCode <= 599);
+    }
+
+    private void retryResponse(CloseableHttpResponse response, int retryIndex) {
+        if (response == null) {
+            return;
+        }
+
+        int statusCode = response.getStatusLine().getStatusCode();
+        int sleepMilliSeconds = RETRY_SLOTS[retryIndex] * 1000;
+        try {
+            LOGGER.info("Will retry for status code {} after {} milli-seconds", statusCode, sleepMilliSeconds);
+            Thread.sleep(sleepMilliSeconds);
+        } catch (InterruptedException e) {
+            throw new IllegalStateException(String.format("Failed to retry for response %s with status " +
+                    "code %s.", toString(response), statusCode), e);
+        }
+    }
+
+
+    private MSIType checkMSIType() {
+        // Two types of MSI credential supported, TODO (w) may split implementation into different files
+        if (StringUtils.hasText(System.getenv("MSI_ENDPOINT")) &&
+                StringUtils.hasText(System.getenv("MSI_SECRET"))) {
+            return MSIType.APP_SERVICE;
+        } else {
+            return MSIType.VM;
+        }
+    }
+
+    private AzureInstanceMetadataService getMetadataService(MSIType type) {
+        switch (type) {
+            case VM:
+                return new AzureInstanceMetadataService();
+            case APP_SERVICE:
+                return new AzureInstanceMetadataService().withTokenEndpoint(System.getenv("MSI_ENDPOINT"));
+            default:
+                return new AzureInstanceMetadataService();
+        }
+    }
+
+    private void setHeader(MSIType type) {
+        switch (type) {
+            case VM:
+                tokenReqHeaders.putAll(ImmutableMap.of("Metadata", "true"));
+                break;
+            case APP_SERVICE:
+                tokenReqHeaders.putAll(ImmutableMap.of("Secret", System.getenv("MSI_SECRET")));
```

Comment:
Can we reuse the map object here? Didn't see it will be changed after initialization.